src/util/strencodings.cpp

```diff
@@ -81,16 +81,21 @@ template <typename Byte>
 std::optional<std::vector<Byte>> TryParseHex(std::string_view str)
 {
     std::vector<Byte> vch;
-    auto it = str.begin();
-    while (it != str.end()) {
-        if (IsSpace(*it)) {
```

Comment:
we don't expect most hex strings to contain spaces so let's fall back to this when we can't parse the string instead

Reply:
Valid question, that's why I put this change in a separate commit with before/after measurements, resulting in an additional 30% speedup on my machine:
[Before:](https://github.com/bitcoin/bitcoin/pull/29458/commits/e467437dea8b78e87ba16af306304c7740d5815c)
```
|           ns/base16 |            base16/s |    err% |     total | benchmark
|--------------------:|--------------------:|--------:|----------:|:----------
|                0.69 |    1,445,181,075.98 |    0.5% |      0.01 | `TryParseHexBenchmark`
|                0.68 |    1,459,941,407.03 |    0.2% |      0.01 | `TryParseHexBenchmark`
|                0.69 |    1,457,349,650.60 |    1.1% |      0.01 | `TryParseHexBenchmark`
```
[After:](https://github.com/bitcoin/bitcoin/pull/29458/commits/a37a5b11bfba8e8cd2c5f9e940770b6b844ecedc)
```
|           ns/base16 |            base16/s |    err% |     total | benchmark
|--------------------:|--------------------:|--------:|----------:|:----------
|                0.54 |    1,866,216,380.08 |    0.9% |      0.01 | `TryParseHexBenchmark`
|                0.53 |    1,871,755,872.12 |    0.5% |      0.01 | `TryParseHexBenchmark`
|                0.53 |    1,875,106,332.14 |    0.1% |      0.01 | `TryParseHexBenchmark`
```

Reply:
Since I'm here, tried this as well. On my machine the speedup is not so severe, but still measurable:

Preallocation only (e467437dea8b78e87ba16af306304c7740d5815c)
|           ns/base16 |            base16/s |    err% |      ins/base16 |      cyc/base16 |    IPC |     bra/base16 |   miss% |     total | benchmark
|--------------------:|--------------------:|--------:|----------------:|----------------:|-------:|---------------:|--------:|----------:|:----------
|                0.84 |    1,186,108,418.00 |    0.3% |           15.21 |            1.86 |  8.172 |           3.37 |    0.0% |      0.53 | `TryParseHexBenchmark`

Skipping whitespace check as well (a37a5b11bfba8e8cd2c5f9e940770b6b844ecedc)
|           ns/base16 |            base16/s |    err% |      ins/base16 |      cyc/base16 |    IPC |     bra/base16 |   miss% |     total | benchmark
|--------------------:|--------------------:|--------:|----------------:|----------------:|-------:|---------------:|--------:|----------:|:----------
|                0.76 |    1,320,747,604.66 |    0.1% |           14.24 |            1.67 |  8.521 |           2.87 |    0.0% |      0.54 | `TryParseHexBenchmark`

Reply:
Hmm thinking about this more, maybe saving 0.08ns is not worth this churn in the codebase. I think I agree with @maflcko that this isn't necessary and preallocation on its own is a simpler change with bigger benefit.

Reply:
Thanks guys, the code is indeed cleaner without this, I've rebased the PR, please check it again

Reply:
Since we've left the `IsSpace` calls here (which were checking the ASCII whitespaces one-by-one for each converted character), I've optimized it in a separate PR to short circuit for non-whitespace cases, see: https://github.com/bitcoin/bitcoin/pull/29602